My Jetpack: Add "Expired" & "Expires soon" statuses to product cards

elliottprogrammer opened the pull request:
This PR adds an "Expired" or "Expiring soon" status to the product cards of products with an expired or expiring paid plan subscription. We also make the product CTA say "Renew my plan" (or "Resume my plan" if expired) and link to the specific subscription in purchase management section of their wordpress.com account so that they can renew or resume their subscription.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

### Screenshots:

![Screen Shot 2024-11-26 at 08 12 58](https://github.com/user-attachments/assets/090e4024-a67d-4e56-9143-261a4b0b4d6d)

![Screen Shot 2024-10-22 at 18 57 57](https://github.com/user-attachments/assets/6131c4f6-df0f-40a2-aeb0-4e02cf5c05f5)

.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Added an "Expired" or "Expiring soon" status to the My Jetpack product cards.
* For products that are Expired, the CTA says "Resume my plan" and points to the wordpress.com purchase management page. For products that are Expiring soon, the CTA says "Renew my plan", and points directly to checkout with the product renewal in the cart.
* Also fixes issue: [Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner](https://github.com/Automattic/jetpack/issues/39762). - Testing instructions below ⤵️

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-bJN-p2

Design Thread: p1HpG7-umC-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
**To test this PR, skip down to "Testing this PR":**
Otherwise, if testing issue: _Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner_, follow the instructions below:


#### Testing issue https://github.com/Automattic/jetpack/issues/39762 :<br/>(Backup Card: My Jetpack shows "Needs Plan" for backup even when there is a Jetpack plan from a hosting partner.)
- Create a Pressable hosting partner site (if you don't already have one) by following instructions here: PCYsg-9TF-p2
    - (sign up using the direct link to the a8c employee plan that requires no coupon)
- Install the Jetpack Beta Tester plugin: https://github.com/Automattic/jetpack-beta
- Activate this PR branch for Jetpack in Jetpack Beta plugin.
- Purchase a Backup plan.
- Open My Jetpack and confirm that Backup product card status is Active and not "Needs plan". 

#### Testing this PR:
* Test this PR with a Jurassic Ninja Site running the Beta plugin on the `add/mj-expired-products-cards` branch. In the JN interface, you can automatically assign paid product licenses for all the products shown on the My Jetpack page, See screenshot:

![Screen Shot on 2024-10-22 at 22-08-12](https://github.com/user-attachments/assets/e041055f-71fd-4650-b46a-8f79025ea6f6)
* Connect Jetpack, both site and user.  (Once user connected, you will see all the paid plans in My Jetpack.)
* Confirm that all paid products are showing the correct status (For example confirm Backup and Search are not showing "Needs plan" when the paid plan has actually been purchased) - See issue: https://github.com/Automattic/jetpack/issues/39762
* Now look up your site in Store Admin. 
* One by one, for each of the subscriptions, change the expiration date to either expiring (nearing expiration), or expired . You may need to disable 'auto-renew' for the plan to show as expiring/expired. The go to My Jetpack, reload the page and look at the product card for that plan.
* Verify the status is either "Expired" or "Expiring soon", depending on the date you set.
* Verify there is a red border around expired products, and a yellow border around products expiring soon.
* Verify that for products that are Expired, the CTA says "Resume my plan" and points to the wordpress.com purchase management page. For products that are Expiring soon, verify the CTA says "Renew my plan", and points directly to checkout with the product renewal in the cart.
* Continue this (set the expiration date, then view in My Jetpack) for each of the products.
* Once done testing all products, in Store Admin you can remove all the product subscriptions.
* Now Purchase a bundle for the site, such as Complete, or Security.
* In Store Admin, set the bundle subscription date to expired/expiring. (you may need to disable auto-renew).
* Now view the products included with the bundle on My Jetpack.  The products included with the bundle should be showing as expired/expiring soon, in My Jetpack.
* Verify that the expired/expiring CTA points to the bundle subscription in the purchase management page on wordpress.com (for expired), or points directly to checkout (for expiring-soon).
* Establish a site-only connection (no user-connection) and verify everything works properly with no errors. Some product cars may now show the status "Requires user connection", which takes precedence over expired/expiring products.

elliottprogrammer committed:
Add expired statuses to products, WIP.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Update & refine logic to determine if products is expired/expiring.

elliottprogrammer committed:
Add logic in remaining product classes for expired/expiring functionality.

elliottprogrammer committed:
Add manage purchase/subscription url for My Jetpack products.

elliottprogrammer committed:
Fix $status possibly null when checking expiry status.

elliottprogrammer committed:
Add error/warning border color to expired/expiring product cards.

IanRamosC commented on projects/packages/my-jetpack/_inc/components/product-card/style.module.scss:
```diff
@@ -5,6 +5,14 @@ $box-shadow-color: rgba( 0, 0, 0, 0.1 );
 .container {
 	position: relative;
 	min-height: 200px;
+	&.has-error {
+		border-width: 2px;
+		border-color: var( --jp-red-50 );
+	}
+	&.has-warning {
+		border-width: 2px;
+		border-color: #DEB100;
+	}
```
I saw that the styles show that the borders are 2px wide, but I personally think that 1px looks better and less aggressive, even though 2px draws more attention to the cards:

2px | 1px
--- | --- 
<img width="368" alt="image" src="https://github.com/user-attachments/assets/b6fe4163-ff29-44e7-b401-8116797bc396">|<img width="368" alt="image" src="https://github.com/user-attachments/assets/37020750-9c4a-43bd-9b19-a1c29b29f007">
<img width="1095" alt="image" src="https://github.com/user-attachments/assets/f4b84b3c-6774-4490-87e8-ccc78857d0b2">|<img width="1095" alt="image" src="https://github.com/user-attachments/assets/0065e946-48da-48a7-a639-3d54ee8493ad">

elliottprogrammer replied:
Ok yeah, sounds good. Thanks Ian, I changed the error & warning border to 1px. 👍

IanRamosC commented on projects/packages/my-jetpack/src/products/class-anti-spam.php:
```diff
@@ -106,6 +106,41 @@ public static function get_features() {
+	 */
+	public static function get_paid_bundles_that_include_product() {
+		return array(
+			'jetpack_security_t1_yearly',
+			'jetpack_security_t1_monthly',
+			'jetpack_security_t1_bi-yearly',
+			'jetpack_complete',
+			'jetpack_complete_monthly',
+			'jetpack_complete_bi-yearly',
+			'jetpack_business',
+			'jetpack_business_monthly',
+			'jetpack_premium',
+			'jetpack_premium_monthly',
+			'jetpack_personal',
+			'jetpack_personal_monthly',
```
Would we like to include WPCOM plans that include Jetpack features or other old Jetpack plans (like Starter) here as well?

IanRamosC replied:
Considering that WPCOM plans would have a different checkout, etc

elliottprogrammer replied:
Yeah, this comment/feedback is no longer relevant. I refactored this so now we're pulling the bundle slugs from the WPCOM features endpoint. I'm marking this comment as "Resolved". 👍

IanRamosC reviewed (dismissed):
This looks pretty good. The changes are easy to follow and the code tested well. 👍🏾

IanRamosC commented:
It seems like you need to update the PR before merging

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-expired-products-cards

elliottprogrammer committed:
Fix how paid products are determined, regarding bundles.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-expired-products-cards

elliottprogrammer committed:
Fix get_site_current_plan() doctype return value.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-expired-products-cards

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-expired-products-cards

elliottprogrammer committed:
Refactor/improve how we determine 'has_paid_plan_for_product()'.

elliottprogrammer committed:
Change product card error & warning border to 1px, per feedback.

elliottprogrammer committed:
Fix red border on non-connected product cards.

CodeyGuyDylan commented on projects/packages/my-jetpack/src/products/class-product.php:
```diff
@@ -379,12 +392,42 @@ public static function get_standalone_info() {
 	 */
 	public static function has_paid_plan_for_product() {
+		// First check site features (if there's a feature that identifies the paid plan)
+		if ( static::$feature_identifying_paid_plan ) {
```
Seems like having this check fulfills most of what [this issue](https://github.com/Automattic/jetpack-marketing/issues/905) needs to be fixed 😄  Once this is done, it should be a quick patch to get this one done finally 🥳

robertsreberski replied:
@grzegorz-cp would you be able to follow up on this, and get [the issue](https://github.com/Automattic/jetpack-marketing/issues/905) done?

CodeyGuyDylan commented on projects/packages/my-jetpack/src/products/class-creator.php:
```diff
@@ -319,24 +319,30 @@ public static function get_wpcom_monthly_product_slug() {
 	}
```
The creator class won't really be doing anything now that Growth is here, so we probably don't need to update it. But I guess it's good for complete-ness

elliottprogrammer replied:
Yeah, I just made them all the same & consistent with any changes I made. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/src/products/class-crm.php:
```diff
@@ -173,6 +173,19 @@ public static function has_paid_plan_for_product() {
+	 *
+	 * @return array
+	 */
+	public static function get_paid_bundles_that_include_product() {
```
Do we only have this function in CRM since CRM exists outside of our system?

elliottprogrammer replied:
Yes, all we can check for CRM is if the site as a Jetpack Complete plan (because CRM is included with Jetpack Complete). That's all we were doing previously, so that's all we are doing here too, so nothing has changed really.

CodeyGuyDylan reviewed (changes requested):
I mentioned this is the [other PR](https://github.com/Automattic/jetpack/pull/40115) but wanted to make sure it didn't get lost. I don't think we should be showing the "expired" status if it's been expired for a very long time. My gut reaction is we probably shouldn't show it if it's been expired for longer than a month

I noticed the "needs connection" error shows as a red error. I'm wondering if it should be yellow since the status is also yellow? Also, if the "Needs connection" error has a red or maybe yellow border, should the "Needs plan" status also have one?

![image](https://github.com/user-attachments/assets/85e573ec-3fd7-4c6e-bc9f-912d512a597c)

Other than those thoughts, the plan recognition upgrade seems to work great! Thank you for your hard work on this

elliottprogrammer committed:
Add option to hide expiration status & and get_renew_url function.

elliottprogrammer commented:
First, thanks very much for the reviews @CodeyGuyDylan! 😃
Oh, my apologies Dylan, I meant to address/answer these points/feedback from your first review.  

> I mentioned this is the [other PR](https://github.com/Automattic/jetpack/pull/40115) but wanted to make sure it didn't get lost. I don't think we should be showing the "expired" status if it's been expired for a very long time. My gut reaction is we probably shouldn't show it if it's been expired for longer than a month

For not showing a product as "expired" after it's been expired for a certain amount of time, here's my thoughts:
If we don't show when the product is expired (after a certain time), then what status do we show it as then? Just "Active"?, or "Needs plan"? For products that require a paid plan to use them, if we show them as Active, it won't actually be active because it's expired. But we show the product as "Needs plan" instead, I don't think the user can re-purchase the same plan until they "Remove" the expired one first.  Something about showing a particular product card (like Backup, or whatever) like nothing is wrong with it, but then when the user clicks on the card to use it, it's not working properly because the plan subscription has been expired. It just seems like it might be opening a big can of worms..
When a product subscription expires, the user really should either 1.- Renew the subscription, Or 2.- Remove it (in the purchase management section). If they Remove the subscription, the expiration status and notice goes away. If they Renew the subscription, the status and notice goes away.  Thats why when a product is expired, the Notice CTA, I pointed to the purchase/subscription management section on wordpress.com where the user can either "Remove" the subscription or "Renew" it. However if the product is expiring soon, the Notice CTA points straight to checkout with the product renewal in the cart.
I noticed that Atomic sites also don't hide the plan expiration notice (after any amount of time) either.  In fact Atomic sites have the expiration notice displaying all over the place, on many different pages all throughout wp-admin (Every one of the settings pages (14 pages), the WP Dashboard page, the Hosting -> home page, the Jetpack - Akismet page, and others I think.), where I believe the thinking is that the user is expected to either renew the subscription, or remove it, if they want the notice to go away. (see screenshots: Here's only a few, out of many, many:)

![Screen Shot 2024-11-25 at 20 16 56](https://github.com/user-attachments/assets/b1b339de-103a-42f7-94d4-789a6c84ce94)
![Screen Shot 2024-11-25 at 20 18 29](https://github.com/user-attachments/assets/bd94c1c0-421d-4a44-aa3d-19fbd8346de5)
![Screen Shot 2024-11-25 at 20 19 50](https://github.com/user-attachments/assets/98aaf4c6-2c2f-419e-b39d-b13e4578404f)

Anyway, so let me know your thoughts. Let me know if maybe I'm missing something, or maybe over-thinking it?

I did however make a change in the code where I added in the ability to hide/don't show the expiration status after 2 months after the expiration date. I didn't turn it on by default. But it would only be a matter of passing in a `true` value into a function call, if we should decide to go that route. 😉🤷‍♂️

> I noticed the "needs connection" error shows as a red error. I'm wondering if it should be yellow since the status is also yellow? Also, if the "Needs connection" error has a red or maybe yellow border, should the "Needs plan" status also have one?

Hmm, I changed/fixed this prior to your review.. I wonder if your branch (or Jetpack Beta branch) was not up to the latest?
I changed the product card border back to how it was originally, where connection warnings and other warnings/errors do not show a red or yellow border around the product cards. Only the expired/expiring products show the red/yellow borders around them. 🙂

CodeyGuyDylan commented:
> Thats why when a product is expired, the Notice CTA, I pointed to the purchase/subscription management section on wordpress.com where the user can either "Remove" the subscription or "Renew" it.

This is a good point and maybe fixes the issue I had with it 🤔 I don't know that, as a user, I would find it clear that the way to make the notification go away is to remove the subscription altogether, though I guess I would probably try it eventually. (I hate notifications 😆). I think we are probably okay to keep it how it is then if that's the case. Could we add to tracks the ability to see how long a subscription has been expired when viewing a card with an expired plan? That way we could check in tracks later to see if users are leaving that notification there for long periods of time or not later on.

> In fact Atomic sites have the expiration notice displaying all over the place

Personally I don't think that's very good UX either 😅. I am not every user but personally that would bug the heck out of me

elliottprogrammer commented:
@CodeyGuyDylan ,

> I don't know that, as a user, I would find it clear that the way to make the notification go away is to remove the subscription altogether, though I guess I would probably try it eventually. (I hate notifications 😆)

Yeah, so true, I think _we all_ hate notifications! 😆

Looking into it and thinking about this further, I'm noticing there's a way to show a close(X) icon on the notice!  I'll look now to see how much it would take to close the expired/expiring notices with some persistence!  This would be ideal, I think.  Do you agree? 😃

> Could we add to tracks the ability to see how long a subscription has been expired when viewing a card with an expired plan? That way we could check in tracks later to see if users are leaving that notification there for long periods of time or not later on.

This is a great idea!  Yes, I'll do this! 🙌

> > In fact Atomic sites have the expiration notice displaying all over the place
> 
> Personally I don't think that's very good UX either 😅. I am not every user but personally that would bug the heck out of me

Yeah, I agree.

CodeyGuyDylan reviewed (dismissed):
More of a question instead of a requested change. In the scenario below, Boost, Social, and Protect all have expired plans. I see that some statuses override the expired plan status such as these. I think that is fine, as it wouldn't work anyway without the plugin being installed or active, so the statuses shown are still integral to getting the plugin to work. I just wanted to make sure it was expected behavior 😄 
![image](https://github.com/user-attachments/assets/a1ad14bc-0707-42c9-b1f9-b73e0c2731ac)

One note, and this probably isn't feasible, but I think it would be great if we could redirect them back to My Jetpack after renewing or removing a product after getting to `/me/purchases` from My Jetpack. Maybe we could do that sometime in the future

Other than that I think all of my issues have been addressed. Thank you for all the changes! I'll put my conversation with payments about the renewal stuff here since that answers the question as to why we don't need to cut off renewals older than a month old 😄 

Slack: p1732730061873749-slack-C02BEP610

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-expired-products-cards

elliottprogrammer committed:
Fix error resulting from merge.

manzoorwanijk commented on projects/packages/my-jetpack/src/products/class-social.php:
```diff
@@ -161,40 +168,42 @@ public static function get_status() {
+	 * @return array
+	 */
+	public static function get_paid_plan_product_slugs() {
+		return array(
+			'jetpack_social_v1_yearly',
+			'jetpack_social_v1_monthly',
+			'jetpack_social_v1_bi_yearly',
+			'jetpack_social_basic_yearly',
+			'jetpack_social_monthly',
+			'jetpack_social_basic_monthly',
+			'jetpack_social_basic_bi_yearly',
+			'jetpack_social_advanced_yearly',
+			'jetpack_social_advanced_monthly',
+			'jetpack_social_advanced_bi_yearly',
+		);
```
Should this contain all the legacy products that are no longer used anymore?

Social currently has only those `*_v1_*` products.

elliottprogrammer replied:
Hey @manzoorwanijk! 👋  Thanks for the question! 😃
> Should this contain all the legacy products that are no longer used anymore?
> Social currently has only those *_v1_* products.

To answer your question: Yes, it should contain ALL possible Social product slugs (including legacy products, but not including Bundle plans)
Yes, ultimately `has_paid_plan_for_product()` uses this function in question (`get_paid_plan_product_slugs()`) as well as `get_paid_bundles_that_include_product()` to get All the paid plan product_slugs and bundle plan product_slugs that include Social, to ultimately determine if the site has a paid plan for the product (in this case Social).
Some sites/users might still be renewing a legacy Social product that they bought previously, so we need to know if the site/user has _ANY_ Social paid plan (including legacy plans, or even bundled plans that include Social within it). Basically, if the site/user has _Any_ paid plan that includes Social, we want to know about it. 😉

manzoorwanijk replied:
Hey

Oh yeah. Thanks.
